### PR TITLE
🐛 HubSpot source: fix CRMSearchStream when incremental update is larger than 10k

### PR DESCRIPTION
**Pull request by TSkrebe**
## What
HubSpot search API which is used during incremental runs of company, contacts, etc. streams, [has a limitation of 10k results](https://developers.hubspot.com/docs/api/crm/search#limitations). Therefore, increments with more than 10k changes fail with `400 Client Error: Bad Request for url: `

## How
1. Return search results is ascending order
2. If we are about to cross 10k threshold, then change request payload by:
   1. reseting "after" to 0
   2. replacing existing starting point (last_modified_field) with newest returned value.

## 🚨 User Impact 🚨
Should be none

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 1]
```diff
@@ -589,31 +593,39 @@ def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
...
             }
             if self.state
             else {}
         )
 
-        while True:
-            stream_records = {}
-            if self.state:
+        after = 0
+        if self.state:
```
Comment (TSkrebe) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:605]: the logic for stateless and stateful paths start to diverge quite a bit and doesn't feel very pythonic. Open to suggestion how to improve this.

**Comment by TSkrebe** [comment]
Curious if it's feasible to simulate 10k in integration tests. it takes less than 1min to load 10k results.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 2]
```diff
@@ -589,31 +593,39 @@ def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
...
+            while after != None:
                 response = getter(data=payload)
                 for record in self._transform(self.parse_response(response)):
-                    stream_records[record["id"]] = record
-            else:
+                    yield record
+                    latest_cursor = self._field_to_datetime(record[self.updated_at_field])
+
+                after = response.get("paging", {}).get("next", {}).get("after")
+                if after and after + self.limit > self.max_search_resuls:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:613]: If `after` is none this won't throw a runtime exception when trying to sum `after + self.limit`?

**Comment by marcosmarxm** [comment]
> Curious if it's feasible to simulate 10k in integration tests. it takes less than 1min to load 10k results.

Out integration test account doesn't have 10k records. Maybe is something feasible to simulate using mock with unit test.

**Comment by marcosmarxm** [comment]
@TSkrebe the response error is handle with this new logic?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py (hunk 2 above)
Comment (TSkrebe) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:613]: we check if after exists so it should all be good:
```
>>> after = None
>>> if after and after + 1 > 0:
...     print("one")
...
>>>
```

will add some unit tests

**Comment by TSkrebe** [comment]
> the response error is handle with this new logic?
can you clarify? there shouldn't be any error if everything works as written in description.

**Comment by TSkrebe** [comment]
@marcosmarxm could you run integration tests to see if everything works as expected?

**Comment by alafanechere** [comment]
Hey @TSkrebe , sorry for the late reply! Could you please fix the conflicts as other PRs were merged for this connector? I'll then run the integration tests!

**Comment by TSkrebe** [comment]
Hi @alafanechere, I see there were some major changes in hubspot connector. Unfortunately, i'll have to defer this PR as I don't have time atm to work on this.

**Comment by marcosmarxm** [comment]
@TSkrebe sorry the long delay to review it. There is another PR open https://github.com/airbytehq/airbyte/pull/10216 solving the same issue but already implemented using the new version of Hubspot Connector. I'll close your PR in favor of the other.
So much thanks for the contribution!

**Pull request closed without merging**